h/views/badge.py

```diff
@@ -21,6 +20,29 @@ def _has_uri_ever_been_annotated(db, uri):
     return result[0] is True
 
 
+class Blocklist:
+    """Block URLs which we know are not worth replying to for the badge."""
+
+    BLOCKED_URL_PARTS = (
+        "//facebook.com",
+        "//mail.google.com",
+        "//www.facebook.com",
+    )
+
+    @classmethod
+    def is_blocked(cls, url):
+        """Check if a URL is blocked."""
+
+        # Dumb is fast here. I can't find a better way of doing this for now
+        url = url.lower()
```

Comment:
Compared to all the machinery (Pyramid, webob etc.) that happens in a no-op web request, more "proper" (in a very loose sense) parsing using urllib might still be very cheap.

Reply:
It's a really significant difference in speed. The test is there to help with experiments like that, so I recommend testing it. It's true that it's small overall, but this works functionally identically to the previous blocklist which used 'LIKE %//domain%'

Some other things I tried which were all quite a bit slower:

 * Using a compound regex to test all at once
 * Url parsing as suggested
 * Trimming the string to the max length of the matching string + 6 (https:) before comparison
 * Prefix matching with startswith

We're splitting hairs because they are all quite quick

Reply:
I've run some experiments with both.

**Current solution:**

Calls per second: 1784121, 0.561 μs/call
Calls per second: 2216312, 0.451 μs/call
Calls per second: 1996805, 0.501 μs/call

**Using urllib parse to get the domain and compare to a set:**

Calls per second: 148918, 6.715 μs/call
Calls per second: 150884, 6.628 μs/call
Calls per second: 164598, 6.075 μs/call

For comparison the current Postgres method is ~200 μs/call. So the original is 400x, and the urllib version is 33x faster.

So there's a bit of variation there, but it's both an order of magnitude slower and still very fast (just not the nanoseconds fast from before). I think it might be worth switching over as both the existing solution and this one would block:

`http://totally.innocent.com/path//facebook.com` 

Even though they shouldn't

Reply:
I've decided on a regex based solution which parses the URL more thoroughly, and retains a similar performance to the dumb solution we started off with.

Previously I discounted this as:

 * It's a tiny bit slower
 * It's more complicated

The advantage is it checks the domain is actually the domain, rather than the pattern just existing anywhere in the string.